### PR TITLE
Feature tests as GitHub actions

### PR DESCRIPTION
**Pull request by felixsch**
## Description

Bring feature tests into Github and run them as Github actions, rather then relying on our internal CI pipelines with its complicated setup.

This change will:
- Run feature tests in github actions
- Auto build the ci-container on master push
- Add scriptlets to easily build rpms and run feature tests

### Run the pipeline locally

- requires you to run a mariadb on your host machine
- Checkout of the branch into <RMT>

```
$ cd <RMT>/ci/
$ docker build --no-cache -t rmt-devel -f Dockerfile .
$ docker run --rm -it -e SCC_USERNAME="<username>" -e SCC_PASSWORD="<password>" --network=host -v $(pwd):/usr/src/rmt-server rmt-devel

> ./ci/rmt-build-rpm 
# expect: in /usr/src/rmt-server/tmp/artifacts/ are built rpms

> ./ci/rmt-configure
# expect: installs and configures rmt from the rpm

> ./ci/rmt-run-feature-tests
# expect: runs the feature tests
```

### CODE REVIEW

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356839/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 9a56c1e5add72b85e75b4d8ab5242e4429dde7fe into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356840/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged bc44b2a7476fd65f4ce5ffa7efd3e75a28998ce2 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356863/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 99449b8b9d921be3498a4b6b30d5535f3a9a6f86 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

File: ci/rmt-configure [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+SOURCE=${SOURCE:-/usr/src/rmt-server}
+
+# Where is the rpm build environment to be found
+BUILD_DIR=${BUILD_DIR:-/usr/src/packages}
+
+# Where copy the built rpm files after a successful build
+ARTIFACT_DIR=${ARTIFACT_DIR:-$SOURCE/tmp/artifacts}
+
+# Current version we are working with
+VERSION=$(ruby -e "require '$SOURCE/lib/rmt.rb'; print RMT::VERSION")
```
Comment (josegomezr) [inline ci/rmt-configure:14]:
```suggestion
VERSION=$(ruby -r "$SOURCE/lib/rmt" -e 'print RMT::VERSION')
```

**Review by josegomezr — COMMENTED** [review]
Have a look at the comments.

and:

- would it make sense to have central bash file with the functions and common prelude (`SOURCE=...`) and then just `source my-prelude.sh`?
- Given that there are more bash scripts, would it make sense to pull `shelllint` into the actions for some sanity checks?

File: .github/workflows/features.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+          --health-timeout=5s
+          --health-retries=3
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: move source to /usr/src/rmt-server
+        run: |
+          [ -d $SOURCE ] && rm -r $SOURCE
```
[the remark this answers is not in the document]
Comment (felixsch) [reply .github/workflows/features.yml:49]: it creates the directory to let it mount later. I'm not sure if this is actually required.

File: .github/workflows/features.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+name: feature tests
+
+on:
+  pull_request:
+    branches: [master]
+
+env:
+  SOURCE: /usr/src/rmt-server
+  SCC_USERNAME: ${{ secrets.SCC_USERNAME }}
+  SCC_PASSWORD: ${{ secrets.SCC_PASSWORD }}
```
[the remark this answers is not in the document]
Comment (felixsch) [reply .github/workflows/features.yml:11]: I checked. Secrets are not exposed when running actions from forks.
vvv from the secrets page.
image

I will look into narrowing it down further

File: Makefile [hunk 4]
```diff
@@ -10,7 +10,7 @@ clean:
 	rm -rf $(NAME)-$(VERSION)/
 
 man:
-	bundle exec ronn --roff --pipe --manual RMT MANUAL.md > rmt-cli.8 && gzip -f rmt-cli.8
+	ronn --roff --pipe --manual RMT MANUAL.md > rmt-cli.8 && gzip -f rmt-cli.8
```
[the remark this answers is not in the document]
Comment (felixsch) [reply Makefile:13]: this is installed via rpm package. Idea was/is to not have to run `bundle install` when `make dist` Do you think this is a problem?

File: .github/workflows/features.yml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+          --health-retries=3
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: move source to /usr/src/rmt-server
+        run: |
+          [ -d $SOURCE ] && rm -r $SOURCE
+          cp -r $GITHUB_WORKSPACE $SOURCE
```
[the remark this answers is not in the document]
Comment (felixsch) [reply .github/workflows/features.yml:50]: Don't get me started on this stuff. There are multiple bug reports open against `checkout` which does not allow to alter the checkout path outside of original $GITHUB_WORKSPACE. They even suggest to move the result until the fix is provided.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (josegomezr) [reply Makefile:13]: ah, so it comes baked in the base container? In that case: even better, we can resolve this

File: .github/workflows/features.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (josegomezr) [reply .github/workflows/features.yml:50]: I imagine :joy: I'm marking this as resolved, the current state works

File: .github/workflows/features.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+          MYSQL_ROOT_PASSWORD: root
+        ports:
+          - 3306:3306
+        options: >-
+          --health-cmd="mysqladmin ping"
+          --health-interval=10s
+          --health-timeout=5s
+          --health-retries=3
+    steps:
+      - uses: actions/checkout@v2
```
Comment (josegomezr) [inline .github/workflows/features.yml:43]:
```suggestion
      - uses: actions/checkout@v4
```

we're in v4 rn

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356886/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 2ba89d1f4fc2722c44ac9e428414514cccfe7163 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356888/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 212be2268402ab93861572044d87a1d61185436c into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356897/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 196e4f457c0de5f3031358615ec6d50c94b78db1 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

File: .github/workflows/features.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixsch) [reply .github/workflows/features.yml:25]: I added a comment paragraph explaining the current situation. Do you think we need more hardening? @josegomezr

File: .github/workflows/features.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixsch) [reply .github/workflows/features.yml:63]: I let this stay for now but will reiterate when the whole ci container pipeline is up and running!

File: .github/workflows/features.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (josegomezr) [reply .github/workflows/features.yml:25]: I don't know for sure. But we can definitely add it as we discover it

**Comment by felixsch** [comment]
I had a discussion about the *prelude* for the scripts. I decided against it because I favor self-contained scripts over configuration repition. @josegomezr offered to make a follow up PR if he deems it necessary to make the prelude happen.

I added the shell linting and fixed bash scripts in `ci/*`

File: .github/workflows/features.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixsch) [reply .github/workflows/features.yml:25]: of course. I will let the current configuration as it is and we adapt accordingly

**Review by ngetahun — APPROVED** [review]
LGTM!

**Review by josegomezr — APPROVED** [review]
LGTM!

**Pull request merged**